### PR TITLE
[Feature] Support varchar type for L0 of persistent index.

### LINKED ISSUE #9493: [Feature] Support varchar type for L0 of persistent index.
## Feature request

**Is your feature request related to a problem? Please describe.**

persistent index do not support varchar key type.

### PR DESCRIPTION
**Pull request by Linkerist**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9493

### CODE REVIEW

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -922,11 +1354,561 @@ StatusOr<std::unique_ptr<MutableIndex>> MutableIndex::create(size_t key_size, co
...
+        data->set_size(snapshot_size);
+        meta->set_format_version(PERSISTENT_INDEX_VERSION_2);
+        _offset = snapshot_size;
+        std::fill(_page_sizes.begin(), _page_sizes.end(), 0);
+        break;
+    }
+    case kAppendWAL: {
+        for (const auto page_size : _page_sizes) {
+            if (page_size != 0) {
+                IndexWalMetaPB* wal_pb = meta->add_wals();
```
Comment (sevev) [inline be/src/storage/persistent_index.cpp:1824]: I think we should only generate one WALPB of one version. And this wal page contains the contents of all shard in L0

File: be/src/storage/persistent_index.h [hunk 2]
```diff
@@ -116,63 +135,151 @@ class MutableIndex {
...
+private:
+    constexpr static size_t kSliceMaxFixLength = 64;
+    uint32_t _fixed_key_size = -1;
+    uint64_t _offset = 0;
+    std::string _path;
+    std::unique_ptr<WritableFile> _index_file;
+    std::shared_ptr<FileSystem> _fs;
+    std::vector<std::unique_ptr<MutableIndex>> _shards;
+    std::vector<uint64_t> _page_sizes;
+    std::vector<size_t> _sizes;
```
Comment (sevev) [inline be/src/storage/persistent_index.h:280]: Maybe we only need one `_page_size` and `_size` but not an array. Just update the `_page_size` when append wal of shard.

File: be/src/storage/persistent_index.cpp [hunk 3]
```diff
@@ -922,11 +1354,561 @@ StatusOr<std::unique_ptr<MutableIndex>> MutableIndex::create(size_t key_size, co
...
+        break;
+    }
+    case kSnapshot: {
+        std::string file_name = get_l0_index_file_name(_path, version);
+        // be maybe crash after create index file during last commit
+        // so we delete expired index file first to make sure no garbage left
+        FileSystem::Default()->delete_file(file_name);
+        size_t snapshot_size = dump_bound();
+        phmap::BinaryOutputArchive ar_out(file_name.data());
+        if (!dump(ar_out)) {
```
Comment (sevev) [inline be/src/storage/persistent_index.cpp:1801]: I think a snapshot of `ShardByLengthMutableIndex` is a  snapshot set of meta, and we should store the snapshot set continuously in L0 file.

File: be/src/storage/persistent_index.cpp [hunk 4]
```diff
@@ -1982,76 +2737,82 @@ static Status merge_shard_kvs(size_t key_size, std::vector<KVRef>& l0_kvs, std::
...
-            l0_kvs_by_shard[cur_shard_idx].shrink_to_fit();
-            l1_kvs_by_shard[cur_shard_idx].clear();
-            l1_kvs_by_shard[cur_shard_idx].shrink_to_fit();
-            cur_shard_idx++;
-            index_shards_idx = 0;
+    auto writer = std::make_unique<ImmutableIndexWriter>();
+    RETURN_IF_ERROR(writer->init(_path, _version));
+    for (const auto& [key_size, shard_info] : _l0->_shard_info_by_key_size) {
+        const auto& [shard_offset, shard_size] = shard_info;
+        auto size = _l0->_shards[shard_offset]->size();
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/persistent_index.cpp:2759]: this is what you're talking about

File: be/src/storage/persistent_index.cpp (hunk 3 above)
Comment (Linkerist) [reply be/src/storage/persistent_index.cpp:1603]: already does so, this dump is this->dump, not MutableIndex's dump.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/persistent_index.cpp [hunk 5]
```diff
@@ -843,127 +824,382 @@ class FixedMutableIndex : public MutableIndex {
...
+        Slice keys[nums];
+        std::vector<IndexValue> values;
+        values.reserve(nums);
+        for (int i = 0; i < nums; i++) {
+            raw::stl_string_resize_uninitialized(&buff, sizeof(uint32_t));
+            RETURN_IF_ERROR(file->read_at_fully(offset, buff.data(), buff.size()));
+            offset += sizeof(uint32_t);
+            uint32_t kv_pair_size = UNALIGNED_LOAD32(buff.data());
+            raw::stl_string_resize_uninitialized(&buff, kv_pair_size);
+            RETURN_IF_ERROR(file->read_at_fully(offset, buff.data(), buff.size()));
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/persistent_index.cpp:1158]: yes, i think so too, i'll do it in next PR since it do not affact function.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
run starrocks_admit_test

**Pull request merged**